dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java

```diff
@@ -131,14 +140,28 @@ public ProgramRuleEngine( ProgramRuleEntityMapperService programRuleEntityMapper
     {
         List<RuleEffect> ruleEffects = new ArrayList<>();
 
+        String programStageUid = programStageInstance != null ? programStageInstance.getProgramStage().getUid() : null;
+
+        List<ProgramRule> programRules = implementableRuleService
+            .getImplementableRules( program )
```

Comment:
Could we push the program stage filtering down to the JPA/database query layer?

Reply:
Done